chain/store/store.go

```diff
@@ -426,39 +409,85 @@ func (cs *ChainStore) MaybeTakeHeavierTipSet(ctx context.Context, ts *types.TipS
 
 	defer cs.heaviestLk.Unlock()
 
-	if ts.Equals(cs.heaviest) {
-		return nil
+	heaviestWeight, err := cs.weight(ctx, cs.StateBlockstore(), cs.heaviest)
+	if err != nil {
+		return xerrors.Errorf("failed to calculate currentHeaviest's weight: %w", err)
 	}
 
-	w, err := cs.weight(ctx, cs.StateBlockstore(), ts)
-	if err != nil {
-		return err
+	heaviestHeight := abi.ChainEpoch(0)
+	if cs.heaviest != nil {
+		heaviestHeight = cs.heaviest.Height()
 	}
-	heaviestW, err := cs.weight(ctx, cs.StateBlockstore(), cs.heaviest)
+
+	// Before we look at newTs, let's refresh best tipset at current head's height -- this is done to detect equivocation
+	newHeaviest, newHeaviestWeight, err := cs.FormHeaviestTipSetForHeight(ctx, heaviestHeight)
 	if err != nil {
-		return err
+		return xerrors.Errorf("failed to reform head at same height: %w", err)
 	}
 
-	heavier := w.GreaterThan(heaviestW)
-	if w.Equals(heaviestW) && !ts.Equals(cs.heaviest) {
-		log.Errorw("weight draw", "currTs", cs.heaviest, "ts", ts)
-		heavier = breakWeightTie(ts, cs.heaviest)
-	}
+	// Equivocation has occurred! We need a new head NOW!
+	if newHeaviest == nil || newHeaviestWeight.LessThan(heaviestWeight) {
+		log.Warnf("chainstore heaviest tipset's weight SHRANK from %d (%s) to %d (%s) due to equivocation", heaviestWeight, cs.heaviest, newHeaviestWeight, newHeaviest)
+		// refresh heaviestWeight 10 times moving up and down
+
+		for i := heaviestHeight + 5; i > heaviestHeight-5; i-- {
+			possibleHeaviestTs, possibleHeaviestWeight, err := cs.FormHeaviestTipSetForHeight(ctx, i)
+			if err != nil {
+				return xerrors.Errorf("failed to produce head at height %d: %w", i, err)
+			}
 
-	if heavier {
-		// TODO: don't do this for initial sync. Now that we don't have a
-		// difference between 'bootstrap sync' and 'caught up' sync, we need
-		// some other heuristic.
+			if possibleHeaviestWeight.GreaterThan(newHeaviestWeight) {
+				newHeaviestWeight = possibleHeaviestWeight
+				newHeaviest = possibleHeaviestTs
+			}
+		}
+
+		if newHeaviest == nil {
+			return xerrors.Errorf("failed to refresh to a new valid tipset")
+		}
 
-		exceeds, err := cs.exceedsForkLength(ctx, cs.heaviest, ts)
+		errTake := cs.takeHeaviestTipSet(ctx, newHeaviest)
+		if errTake != nil {
+			return xerrors.Errorf("failed to take newHeaviest tipset as head: %w", err)
+		}
+	}
+
+	// if the new height we were notified about isn't what we just refreshed at, see if we have a heavier tipset there
+	if newTsHeight != newHeaviest.Height() {
```

Comment:
I'm not sure how this is possible? If we refresh at a different height, shouldn't we have picked the heaviest tipset at that height?

Reply:
No, not necessarily. We'll only refresh at different heights if there was equivocation. Basically, assume current height `heaviestHeight` is 10, and we were informed about `newTsHeight` of 11.

- We refresh by finding the heaviest tipset at 10
- If the "refreshed" tipset at 10 is not lighter than our current heaviest tipset at 10, we'll just move on
- At that point, we hit this line, and decide we should form the heaviest tipset at height 11, and compare the two.

(If the "refreshed" tipset WAS lighter, though, then yes we would likely form a heavier tipset at 11, and immediately take that as our head, in which case we'd short-circuit here)

Reply:
I think I understand?

Reply:
Let's talk through tomorrow, but I'm pretty sure it's correct.